### PR TITLE
Addressing `Pipeline` edge cases and expanding tests.

### PR DESCRIPTION
**Pull request by WillJRoper**
This PR addresses some issues observed by Xiaowei Ou with the Pipeline. These include:

- The inability to pass cosmology objects to `get_images_luminosity` meaning the internal conversion between cartesian and angular coordinates can't be done.
- No clear feedback on noise units.
- The inability to write interim images out.

This PR addresses these issues by:

- Adding the cosmo argument to `get_images_luminosity`.
- Adding a validation stage to clearly signpost when an instrument is incompatible. 
- Adding an argument to signal that all images should be written.

In an attempt to catch similar issue, the test suite has been expanded to 100% coverage for the pipeline module.

## Issue Type

- Bug
- Enhancement

## Checklist
- [x] I have read the [CONTRIBUTING.md]() -->
- [x] I have added docstrings to all methods
- [x] I have added sufficient comments to all lines
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no pep8 errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

* **New Features**
  * Optional cosmology parameter added to image-generation paths to support angular calculations.
  * write_all flag added to control whether intermediate images (base/PSF/noise) are retained or discarded.
  * Runtime validation added to ensure instrument noise depth/maps use compatible units with clear error messages.

* **Documentation**
  * Public docstrings updated to describe the new parameters and behavior.

* **Tests**
  * Expanded tests covering cosmology propagation, write_all behavior, and noise-unit validation.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tests/test_pipeline.py [hunk 1]
```diff
@@ -909,3 +909,760 @@ def test_run_pipeline_sfh(
...
+
+        # Create instrument with correct units (test that no error is raised)
+        _inst = Instrument(
+            label="test_inst",
+            filters=None,
+            resolution=1.0 * kpc,
+            noise_maps=noise_map,
+        )
+
+        # Should not raise
```
[the remark this answers is not in the document]
Comment (tHarvey303) [reply tests/test_pipeline.py]: Unless I'm missing something I think the rabbit has a point here - we want to run `validate_noise_unit_compatibility` for these 4 tests as well otherwise they are only instantiating a instument and not checking the noise units.

**Review by tHarvey303 — CHANGES_REQUESTED** [review]
Generally looks good, one question on the tests and one very minor edge case on the noise/depths types. Tests run fine locally so should be good to go after this.

**Review by tHarvey303 — APPROVED** [review]
Looks good, apparent magnitudes now work and tests actually test all cases.
